backends/hip/ceed-hip-vector.c

```diff
@@ -0,0 +1,397 @@
+  case CEED_OWN_POINTER:
+    ierr = hipFree(data->d_array_allocated); CeedChk_Hip(ceed, ierr);
+    data->d_array_allocated = array;
+    data->d_array = array;
+    break;
+  case CEED_USE_POINTER:
+    ierr = hipFree(data->d_array_allocated); CeedChk_Hip(ceed, ierr);
+    data->d_array_allocated = NULL;
+    data->d_array = array;
+    break;
+  }
+  data->memState = CEED_HIP_DEVICE_SYNC;
+  return 0;
+}
+
+//------------------------------------------------------------------------------
+// Set the array used by a vector,
+//   freeing any previously allocated array if applicable
+//------------------------------------------------------------------------------
+static int CeedVectorSetArray_Hip(const CeedVector vec,
+                                  const CeedMemType mtype,
+                                  const CeedCopyMode cmode,
+                                  CeedScalar *array) {
+  int ierr;
+  Ceed ceed;
+  ierr = CeedVectorGetCeed(vec, &ceed); CeedChk(ierr);
+  CeedVector_Hip *data;
+  ierr = CeedVectorGetData(vec, (void *)&data); CeedChk(ierr);
+
+  switch (mtype) {
+  case CEED_MEM_HOST:
+    return CeedVectorSetArrayHost_Hip(vec, cmode, array);
+  case CEED_MEM_DEVICE:
+    return CeedVectorSetArrayDevice_Hip(vec, cmode, array);
+  }
+  return 1;
+}
+
+//------------------------------------------------------------------------------
+// Vector Take Array
+//------------------------------------------------------------------------------
+static int CeedVectorTakeArray_Hip(CeedVector vec, CeedMemType mtype,
+                                   CeedScalar **array) {
+  int ierr;
+  CeedVector_Hip *impl;
+  ierr = CeedVectorGetData(vec, (void *)&impl); CeedChk(ierr);
+
+  switch(mtype) {
+  case CEED_MEM_HOST:
+    if (impl->memState == CEED_HIP_DEVICE_SYNC) {
+      ierr = CeedSyncD2H_Hip(vec); CeedChk(ierr);
+    }
+    (*array) = impl->h_array;
+    impl->h_array = NULL;
+    impl->h_array_allocated = NULL;
+    impl->memState = CEED_HIP_HOST_SYNC;
+    break;
+  case CEED_MEM_DEVICE:
+    if (impl->memState == CEED_HIP_HOST_SYNC) {
+      ierr = CeedSyncH2D_Hip(vec); CeedChk(ierr);
+    }
+    (*array) = impl->d_array;
+    impl->d_array = NULL;
+    impl->d_array_allocated = NULL;
+    impl->memState = CEED_HIP_DEVICE_SYNC;
+    break;
+  }
+
+  return 0;
+}
+
+//------------------------------------------------------------------------------
+// Set host array to value
+//------------------------------------------------------------------------------
+static int CeedHostSetValue(CeedScalar *h_array, CeedInt length,
+                            CeedScalar val) {
+  for (int i = 0; i < length; i++)
+    h_array[i] = val;
+  return 0;
+}
+
+//------------------------------------------------------------------------------
+// Set device array to value (impl in .hip file)
+//------------------------------------------------------------------------------
+int CeedDeviceSetValue(CeedScalar *d_array, CeedInt length, CeedScalar val);
```

Comment:
We need to append the suffix `_Hip` and `_Cuda` to these functions (`CeedDeviceSetValue` and `CeedHostSetValue') because the two definitions are conflicting when I build on a machine with both an NVIDIA and AMD card

Reply:
Good catch, fixed in 9550c04208015ba4695cd180d78af9b1f05c6d07